Feature/report user without 2factor

idelcano opened the pull request:
clickup issue: 
https://app.clickup.com/t/8694bw0cq

I have done a significant refactor; it's not finished yet, but we could merge this branch with the permission fixer and two-factor checker user monitoring subscripts to move forward.
This PR comes from this https://github.com/EyeSeeTea/d2-tools/pull/42 but have more funtionallity (two factor checker), and some coding dojo sessions.

Todo in the future:
We should add some tests and refactor some logic that's still in the repositories but should be in the use case. Also, use the d2api typed filters instead of hardcoded endpoints.


How check:
To check this branch, you can get the d2-tools datastore from the dev environment and check the README.
I need to set Node.js to version 16 before building to make the file resources work when pushing to DHIS2. Arnau added this functionality some time ago in his commits.

Commands example:

To check the permission fixer:
yarn start usermonitoring run-permissions-fixer --config-file config.json

To check the missing two-factor:
yarn start usermonitoring run-2fa-reporter --config-file config.json

nshandra commented on src/domain/entities/user-monitoring/permission-fixer/PermissionFixerConfigOptions.ts:
```diff
@@ -0,0 +1,24 @@
+    excludedRolesByUser: RolesByUser[];
+    excludedRolesByGroup: RolesByGroup[];
+    excludedRolesByRole: RolesByRoles[];
+    pushProgramId: NamedRef;
+    minimalGroupId: NamedRef;
+    minimalRoleId: NamedRef;
```
A variable ending with Id is expected to containt just the Id.

nshandra commented on src/domain/usecases/user-monitoring/permission-fixer/RunUserPermissionUseCase.ts:
```diff
@@ -0,0 +1,454 @@
+        pushFixedUserGroups: boolean
+    ): Async<PermissionFixerReport> {
+        if (userIdWithoutGroups != undefined && userIdWithoutGroups.length > 0) {
+            const minimalUserGroup = await this.userGroupRepository.getByIds([minimalGroupId.id]);
```
`minimalUserGroup` should be `PermissionFixerUserGroupExtended`, not `PermissionFixerUserGroupExtended[]`.
This will avoid the `[0]!` hacks below.

nshandra commented on src/domain/usecases/user-monitoring/permission-fixer/RunUserPermissionUseCase.ts:
```diff
@@ -0,0 +1,454 @@
+                            validUserRoles: userValidRoles,
+                            actionRequired: userInvalidRoles.length > 0,
+                            invalidUserRoles: userInvalidRoles,
+                            userNameTemplate: templateGroupMatch!.template.name,
+                            templateIdTemplate: templateGroupMatch!.template.id,
+                            groupIdTemplate: templateGroupMatch!.group.id,
```
To avoid the not undefined asserts you could add a check after creating templateGroupMatch with an exception:
````typescript
if (templateGroupMatch === undefined) {
  throw new TypeError('....');
}
````

nshandra reviewed (changes requested):
Most of the comments are minor, so @idelcano, you can mark some of the lengthy ones for a future refactor.

idelcano commented on src/domain/usecases/user-monitoring/two-factor-monitoring/RunTwoFactorReportUseCase.ts:
```diff
@@ -0,0 +1,44 @@
+            );
+        }
+
+        const usersWithoutTwoFA = usersMustHave2FA.filter(user => {
```
I use TwoFactor to be more consistent with all the names in the code.

idelcano commented on src/utils/uid.ts:
```diff
@@ -0,0 +1,59 @@
+        },
+        { n: hashInteger, uid: "" }
+    );
+
```
I have removed it due is not necessary at this point. (i didnt need use api/system/id either)

idelcano commented on src/domain/usecases/user-monitoring/permission-fixer/RunUserPermissionUseCase.ts:
```diff
@@ -0,0 +1,454 @@
+                .replace(":", "-")
+                .replace("/", "-")
+                .replace("/", "-")
+                .replace("\\", "-");
```
I moved it to UserMonitoringFileResourceUtils

idelcano commented:
Its done. I have implemented a new thing, the list of users sent by 2factor script can't be stored in a dataelement of type long_text, and i have changed it to be a file_resource. In the code i have change te push of that data from text to file. 
And i made common the save file resource things for both scripts, permission fixer and 2factor monitoring

nshandra commented on src/data/user-monitoring/permission-fixer/PermissionFixerUserGroupD2Repository.ts:
```diff
@@ -0,0 +1,45 @@
+            )
+            .getData();
+
+        const userGroups = responses["userGroups"][0] ?? undefined;
+        if (userGroups) {
+            return userGroups;
+        } else {
+            throw new UserGroupNotFoundException("Error getting user groups: " + groupsIds.join(","));
+        }
+    }
```
this should be in `RunUserPermissionUseCase`. `getByIds` usually gets multiple userGroups, hence the array.
As said in the other comment, i think this should become a get with a single id. Because its only used in that way.

nshandra reviewed (approved):
Looks good.
Requested changes fulfilled.

@ifoche Ready to merge.
